src/tactic/doc_commands.lean

```diff
@@ -0,0 +1,523 @@
+It is a finishing tactic, i.e. it is meant to close
+the current goal, not to make some inconclusive progress.
+A mostly trivial example would be:
+
+```lean
+example (a b c : ℕ) (f : ℕ → ℕ) (h: a = b) (h' : b = c) : f a = f c := by cc
+```
+
+As an example requiring some thinking to do by hand, consider:
+
+```lean
+example (f : ℕ → ℕ) (x : ℕ)
+  (H1 : f (f (f x)) = x) (H2 : f (f (f (f (f x)))) = x) :
+  f x = x :=
+by cc
+```
+
+The tactic works by building an equality matching graph. It's a graph where
+the vertices are terms and they are linked by edges if they are known to
+be equal. Once you've added all the equalities in your context, you take
+the transitive closure of the graph and, for each connected component
+(i.e. equivalence class) you can elect a term that will represent the
+whole class and store proofs that the other elements are equal to it.
+You then take the transitive closure of these equalities under the
+congruence lemmas.
+
+The `cc` implementation in Lean does a few more tricks: for example it
+derives `a=b` from `nat.succ a = nat.succ b`, and `nat.succ a !=
+nat.zero` for any `a`.
+
+* The starting reference point is Nelson, Oppen, [Fast decision procedures based on congruence
+closure](http://www.cs.colorado.edu/~bec/courses/csci5535-s09/reading/nelson-oppen-congruence.pdf),
+Journal of the ACM (1980)
+
+* The congruence lemmas for dependent type theory as used in Lean are described in
+[Congruence closure in intensional type theory](https://leanprover.github.io/papers/congr.pdf)
+(de Moura, Selsam IJCAR 2016).
+" }
+
+add_tactic_doc
+{ name := "conv",
+  category := doc_category.tactic,
+  decl_names := [`tactic.interactive.conv],
+  tags := ["core"],
+  description :=
+"`conv {...}` allows the user to perform targeted rewriting on a goal or hypothesis,
+by focusing on particular subexpressions.
+
+See <https://leanprover-community.github.io/mathlib_docs/conv.html> for more details.
+
+Inside `conv` blocks, mathlib currently additionally provides
+* `erw`,
+* `ring` and `ring2`,
+* `norm_num`,
+* `norm_cast`, and
+* `conv` (within another `conv`).
+
+Using `conv` inside a `conv` block allows the user to return to the previous
+state of the outer `conv` block after it is finished. Thus you can continue
+editing an expression without having to start a new `conv` block and re-scoping
+everything. For example:
+```lean
+example (a b c d : ℕ) (h₁ : b = c) (h₂ : a + c = a + d) : a + b = a + d :=
+by conv {
+  to_lhs,
+  conv {
+    congr, skip,
+    rw h₁,
+  },
+  rw h₂,
+}
+```
+Without `conv`, the above example would need to be proved using two successive
+`conv` blocks, each beginning with `to_lhs`.
+
+Also, as a shorthand, `conv_lhs` and `conv_rhs` are provided, so that
+```lean
+example : 0 + 0 = 0 :=
+begin
+  conv_lhs { simp }
+end
+```
+just means
+```lean
+example : 0 + 0 = 0 :=
+begin
+  conv { to_lhs, simp }
+end
+```
+and likewise for `to_rhs`.
+" }
+
+add_tactic_doc
```

Comment:
The CI is failing with `/home/runner/work/mathlib/mathlib/src/tactic/doc_commands.lean:367:0: error: deep recursion was detected at 'infer_type' (potential solution: increase stack space in your system)` [here](https://github.com/leanprover-community/mathlib/runs/499784455#step:5:9). Not sure what I did wrong; is the `description` string too long?

Reply:
Argh, this is annoying. It does seem to come down to the string being "too long." Everything goes fine until line 194 (the `add_decl` line in `add_tactic_doc`). The expression is built and we give it its type explicitly but it looks like the large string causes some kind of overflow in type checking?

Note that we can't even define `def s : string := "..."` with the long doc string.

Reply:
It's not like the string is that long, I don't get what the issue is. (Besides "Lean is really bad at string processing"?)

We could always add it as a doc string to a dummy declaration. Then I'd like to hide that dummy declaration from the doc generation. But it won't work if the declaration is private, and we have to go through the front end so it can't have an `_anonymous` name.

Reply:
> We could always add it as a doc string to a dummy declaration.

Nope, we can't even do that, still a memory error when it tries to use the string.

Reply:
The best workaround I have is to create a definition of the following form:
```
def foo : string :=
let x : list (unit → string) :=
[
λ x, "aaasdsdfsgfsgfsgfdsfgsfsdfsfsfs1\n",
λ x, "aaasdsdfsgfsgfsgfdsfgsfsdfsfsfs2\n",
...
] in
string.join (x.map (λ x, x ()))
```

This should work for relatively large strings.  It is important that each line is in a separate lambda, because there is a VM optimization whose stack usage is linear in the length of a single function...

Reply:
It would be better to add support for string literals to the VM directly.  (This would also help with some performance problems...)

This should be a fairly small project.  We already have an expression macro for string literals.
 1. Add a VM instruction to load string literals.
 2. Write a `find_string_values` function (copy&paste from `find_nat_values`) and hook it up in the preprocessor.
 3. Add two lines in `vm_compiler_fn::compile_macro` to emit the right instruction.

I currently don't really have time to do it myself though.

Reply:
@gebner Thanks for the workaround. It's hard to use in this situation. We have a value of type `tactic_doc_entry` that needs to be reflected to an `expr` and then added to the environment as the body of a definition. `tactic_doc_entry` has a field of type `string` and we want this value to be long. I think we'd need to change the reflection function to produce an expr like the one you show, right?

The other workaround: for now, just leave the simp "tactic doc" in an md file like it was. It's much more verbose than the other tactic entries. It's more like a "simp tutorial" that Kevin wrote. We can add a shorter entry that shows the syntax and options.

Reply:
> It would be better to add support for string literals to the VM directly. (This would also help with some performance problems...)
> 
> This should be a fairly small project. We already have an expression macro for string literals.
> 
> 1. Add a VM instruction to load string literals.
> 2. Write a `find_string_values` function (copy&paste from `find_nat_values`) and hook it up in the preprocessor.
> 3. Add two lines in `vm_compiler_fn::compile_macro` to emit the right instruction.
> 
> I currently don't really have time to do it myself though.

Should we make this an issue on the Lean repo?

Reply:
@robertylewis The workaround is quite easy to use:
```lean
meta def string.to_chunks (size : ℕ) : string → opt_param (list string) [] → list string | s acc :=
if s.length ≤ size then s :: acc else
string.to_chunks (s.popn_back size) (s.backn size :: acc)

section
local attribute [semireducible] reflected
meta instance thunk.reflected {α} {a : thunk α} [has_reflect α] : reflected a :=
expr.lam `x binder_info.default (reflect unit) (reflect (a ()))
end

@[priority 2000]
meta instance string.reflect' (s : string) : reflected s :=
have ts : list (thunk string), from (s.to_chunks 256).map (λ s _, s),
have h : s = string.join (ts.map (λ t, t ())), from undefined,
suffices reflected (string.join $ ts.map (λ t, t ())), by rwa h,
`(string.join $ list.map _ _)

-- now `reflect "foo"` works (until a few hundred kilobytes):
meta def foo : string := by do
have long_string : string,
  from string.join $ (list.range 5000).map (λ _, "x"), do
tactic.exact $ reflect long_string
#eval foo
```

Reply:
> Should we make this an issue on the Lean repo?

https://github.com/leanprover-community/lean/issues/144

Reply:
> The other workaround: for now, just leave the simp "tactic doc" in an md file like it was. It's much more verbose than the other tactic entries. It's more like a "simp tutorial" that Kevin wrote. We can add a shorter entry that shows the syntax and options.

I'm fine with making the doc string for `simp` be its doc entry, and I'll push that fix now. As a reminder to myself, I should go through and turn doc strings into doc entries for the other core tactics, either in this PR or the followup.

Reply:
> @robertylewis The workaround is quite easy to use:

Thanks, this does work. I was able to define a string in two parts like 
```lean
private def foo : string :=
let x : list (unit → string) :=
[
λ x, "aaasdsdfsgfsgfsgfdsfgsfsdfsfsfs1\n",
λ x, "aaasdsdfsgfsgfsgfdsfgsfsdfsfsfs2\n",
...
] in
string.join (x.map (λ x, x ()))
```
and then use `foo` as the `description` of a tactic doc entry. (I still wasn't able to give the full string as the `description` value.)

But since this isn't really essential, maybe we should skip the hack and wait for someone to take up your suggestion adding better support in the VM.

Reply:
> I still wasn't able to give the full string as the description value.

Have you tried `by tactic.exact $ reflect "long string"` (with my reflect instance)?  We could even make notation for it.  But I agree, this is not particularly nice.

Reply:
It's not happy with that.
```lean
add_tactic_doc
{ name := "simp",
  category := doc_category.tactic,
  decl_names := [`tactic.interactive.simp],
  tags := ["core", "simplification"],
  description := by tactic.exact $ string.reflect' /-long string-/ }
```
fails when it tries to elaborate the structure as a `tactic_doc_entry`.

Reply:
Oh yes, of course.  `string.reflect'` is still Lean code.  Unfortunately we don't have a tactic argument parser for strings and I don't see a good way to get strings out of a pexpr.

Well, this leaves improving the VM as the only option if we want longer descriptions.